File: src/main/go/g0001_0100/s0076_minimum_window_substring/solution.go
package s0076_minimum_window_substring

// #Hard #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table #Sliding_Window
// #Level_2_Day_14_Sliding_Window/Two_Pointer #Top_Interview_150_Sliding_Window
// #Big_O_Time_O(s.length())_Space_O(1) #2025_05_06_Time_0_ms_(100.00%)_Space_5.22_MB_(52.57%)

import "math"

func minWindow(s string, t string) string {
	mapT := make([]int, 128)
	for i := 0; i < len(t); i++ {
		mapT[t[i]-'A']++
	}
	count := len(t)
	begin, end, d, head := 0, 0, math.MaxInt32, 0

	for end < len(s) {
		if mapT[s[end]-'A']--; mapT[s[end]-'A'] >= 0 {
			count--
		}
		end++

		for count == 0 {
			if end-begin < d {
				d = end - begin
				head = begin
			}
			if mapT[s[begin]-'A']++; mapT[s[begin]-'A'] > 0 {
				count++
			}
			begin++
		}
	}

	if d == math.MaxInt32 {
		return ""
	}
	return s[head : head+d]
}

File: src/main/js/g0001_0100/s0076_minimum_window_substring/solution.js
// #Hard #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table #Sliding_Window
// #Level_2_Day_14_Sliding_Window/Two_Pointer #Top_Interview_150_Sliding_Window
// #Big_O_Time_O(s.length())_Space_O(1) #2024_12_11_Time_4_ms_(99.86%)_Space_51.7_MB_(98.27%)

/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    const map = new Array(128).fill(0)

    for (let i = 0; i < t.length; i++) {
        map[t.charCodeAt(i)]++
    }

    let count = t.length
    let begin = 0
    let end = 0
    let d = Number.MAX_VALUE
    let head = 0

    while (end < s.length) {
        if (map[s.charCodeAt(end++)]-- > 0) {
            count--
        }

        while (count === 0) {
            if (end - begin < d) {
                d = end - begin
                head = begin
            }
            if (map[s.charCodeAt(begin++)]++ === 0) {
                count++
            }
        }
    }

    return d === Number.MAX_VALUE ? '' : s.substring(head, head + d)
};

export { minWindow }
